app/Main.hs

```diff
@@ -93,62 +100,143 @@ newtype MilliSeconds = MilliSeconds { unMilliSeconds :: Int }
+      <> metavar "FILENAME"
+      <> lookupFromEnv "VAULTENV_SECRETS_FILE"
+      <> help ("Config file specifying which secrets to request. Also configurable " ++
+               "via VAULTENV_SECRETS_FILE." ))
+    <*> argument str
+      (  metavar "CMD"
+      <> help "command to run after fetching secrets")
+    <*> many (argument str
+      (  metavar "ARGS..."
+      <> help "Arguments to pass to CMD, defaults to nothing"))
+    <*>
+      ( flag (esNoConnectTls envSwitches) True
+        (  long "no-connect-tls"
+        <> help ("Don't use TLS when connecting to Vault. Default: use TLS. Also " ++
+                "configurable via VAULTENV_NO_CONNECT_TLS."))
+      <|> flag (esNoConnectTls envSwitches) False
+        (  long "connect-tls"
+        <> help ("Always connect to Vault via TLS. Default: use TLS. Can be used " ++
+                 "to override VAULTENV_NO_CONNECT_TLS.")
+        )
+      )
+    <*>
+      ( flag (esNoValidateCerts envSwitches) True
+        (  long "no-validate-certs"
+        <> help ("Don't validate TLS certificates when connecting to Vault. Default: " ++
+                "validate certs. Also configurable via VAULTENV_NO_VALIDATE_CERTS."))
+      <|> flag (esNoValidateCerts envSwitches) False
+        (  long "validate-certs"
+        <> help ("Always validate TLS certificates when connecting to Vault. Default: " ++
+                 "validate certs. Can be used to override VAULTENV_NO_CONNECT_TLS.")
+        )
+      )
+    <*>
+      ( flag (esNoInheritEnv envSwitches) True
+        (  long "no-inherit-env"
+        <> help ("Don't merge the parent environment with the secrets file. Default: " ++
+                "merge environments. Also configurable via VAULTENV_NO_INHERIT_ENV."))
+      <|> flag (esNoInheritEnv envSwitches) False
+        (  long "inherit-env"
+        <> help ("Always merge the parent environment with the secrets file. Default: " ++
+                 "merge environments. Can be used to override VAULTENV_NO_INHERIT_ENV.")
+        )
+      )
+    <*> (MilliSeconds <$> option auto
+            (  long "retry-base-delay-milliseconds"
+            <> metavar "MILLISECONDS"
+            <> lookupFromEnvWithDefault "VAULTENV_RETRY_BASE_DELAY_MS" 40
+            <> help ("Base delay for vault connection retrying. Defaults to 40ms. " ++
+                     "Also configurable via VAULTENV_RETRY_BASE_DELAY_MS.")))
+    <*> option auto
+      (  long "retry-attempts"
+      <> metavar "NUM"
+      <> lookupFromEnvWithDefault "VAULTENV_RETRY_ATTEMPTS" 9
+      <> help "Maximum number of vault connection retries. Defaults to 9")
   where
-    environ vars key = maybe mempty value (lookup key vars)
+    lookupFromEnv key = foldMap value (lookup key environment)
+    lookupFromEnvWithDefault key defVal = maybe (value defVal) value (readFromEnvironment key)
+
+    readFromEnvironment :: Read a => String -> Maybe a
+    readFromEnvironment var = lookup var environment >>= Read.readMaybe
```

Comment:
This parser has become quite massive. I would break it up into smaller parts:

1. Lift the functions in the `where` clause to the top-level. Add a docstring to them.
2. Put each individual parser on the top-level. Also document them (especially the `flag` and `<|>` story).
3. Combine them with `optionsParser envSwitches environment = Options <$> hostParser <*> portParser ...`
4. Maybe move the parsing functions to a different module?